### ISSUE TITLE
move developer only properties out of UserProfile model

### ISSUE THREAD

**Issue by eviljeff** [mozilla/addons-server#12081]
We have the following properties stored for every user but only need them for the subset of users who have authored an addon because they're displayed on a public facing account profile page that is only accessible for addon authors.  In addition, django also retrieves these extra fields from the database every time the `UserProfile` instance is accessed - which is *a lot*.

We should move them to a seperate model with an FK to UserProfile that's only populated the first time a user sets the properties, and only retrieved for the few occasions we need to show them.

- averagerating
- biography
- homepage
- location
- occupation
- picture_type
- (read_dev_agreement - not strictly a display property, but is developer only.  Might be easier to leave it on UserProfile)

**Comment by Abhi-khandelwal** [comment]
@eviljeff , Can I work on this issue ?

**Comment by eviljeff** [comment]
@Abhi-khandelwal you _could_ but it's not trivial.  You'd need a new model; a database migration to populate it from the existing `UserProfile` model; changes in the account serialiser to get and set the properties when needed; and a number of tests will need changing.

**Comment by Abhi-khandelwal** [comment]
@eviljeff , Can you suggest me the model name ? and also FK field name.

**Comment by eviljeff** [comment]
`UserProfileDetails`?  I guess it should be really be called `UserProfile` but our user model is already called that and it'd be a big change to do that (we can do it later).  The FK could be called `user`.

**Comment by eviljeff** [comment]
(I updated the fields too, to just focus on things that would be set by the developer in a "profile")

**Comment by diox** [comment]
Note that a challenge with this is also making this the extra model is joined every time we _are_ displaying developer information, to avoid extra queries. Because of this I'm not 100% sure it's worth doing.

**Comment by eviljeff** [comment]
It's a fair concern, but I think we only expose those fields via one api endpoint (https://addons-server.readthedocs.io/en/latest/topics/api/accounts.html#account) so it should just be one extra `select_related`.

**Comment by diox** [comment]
True, if we can limit that to fields we're exposing only on the account API endpoint then that's not a problem. That means at the minimum we should leave `picture_type` on `UserProfile` (it's used to build `picture_url`). That would mean the extra model should contain only those fields:
- `biography`
- `homepage`
- `location`
- `occupation`

**Comment by leni1** [comment]
Hi @diox @eviljeff I would like to pick up this issue.

Have you both agreed on the approach or there's still some fleshing out required? 
My next steps would be to review the discussion here, understand the parts of the code that will be affected and come up with a rough implementation that's good enough for review.

Thanks 🙏🏿

**Comment by diox** [comment]
You can give it a shot. The only fields that should be moved to a new models are the ones I listed in my last comment.

**Comment by leni1** [comment]
I took some time to review this and would like to propose the following implementation plan:

1. Make a feature branch for the issue called `12081-move-developer-only-properties-out-of-user-profile-model`. I would keep this up to date with anything on `master`.
2. Make each change in its own branch and merge into the feature branch mentioned above.
3. When done, open a PR for review.

For step 2, I would do:
- Making the extra model with the specified details and then do a database migration to populate [the model]. I would call the model `AddonProfileDetails` since as I understand it, this is for users who've authored add-ons. Also update the model tests accordingly (I think these should be the ones in `users/tests/test_models.py`
- Change the `accounts/serializer.py` as well as associated tests to use the new model.

Question: would `accounts/views.py` also need an update similar to the serializer? 
I would assume not but would like to confirm.

**Comment by diox** [comment]
Your plan sounds good to me, thanks for taking this on!

For step 2, you might not need a branch of each change, but that's up to you, as long as it all gets merged in the feature branch and you make the PR from that branch.

I'd probably avoid calling this `Addon<Something>` just because the models we already have with `Addon` in the name are typically tied directly to a specific addon. Maybe `DeveloperProfileDetails` or `DeveloperDetails` ?

Regarding migrations, one small detail: don't commit a migration removing the old fields just yet - create the ones on the new model and the migration to copy the old data, but leave the old fields for now, so that our database state is backwards-compatible for now. That will make deploying the patch easier for us.

I also don't think you'll need to modify `accounts/views.py`, writes to a related model should be doable from the serializer entirely. Maybe you'll have to tweak some tests for the views though, not sure.

**Comment by leni1** [comment]
Awesome! Will do 👍🏿

**Comment by eviljeff** [comment]
From experience, you will have to do a bit of extra work to save the fields on a different model - rest framework has limitations around saving multiple models at once.

**Comment by KevinMind** [comment]
Old Jira Ticket: https://mozilla-hub.atlassian.net/browse/ADDSRV-16
